### PR TITLE
[SUN-887] Initial joint values are read when the hardware interface starts

### PR DESCRIPTION
**Pull request by tgaspar**
# Description

This PR adds code that reads joint state values from RWS before the hardware interface is started. It is important to note that the code introduced here does not make use of the existing RWS managers that would be available vie `abb_librws` because those are incompatible with RWS2.0. For the sake of not complicating things, I added a simple method that makes the HTTP GET request to the correct address to retrieve the joint values.

Some reading material regarding the issue:
- [Robot arm goes to zero position the moment ros2_control is started](https://github.com/PickNikRobotics/abb_ros2/issues/44)
- [Add support for RWS 2.0](https://github.com/ros-industrial/abb_librws/issues/147)

# Test plan

The test plan revolves mainly around making sure everything builds. The functionality will be tested by me on the real robot. Some additional changes might come into place after the test are done. However, the bulk of the work is here.

# Checklist for Reviewers

Reviewer 1: @fgrcar 
Reviewer 2: @JenniferBuehler 

Before a PR to this repo can get approved and merged, the following needs to be checked by both reviewers:

- The test plan was executed and verified by:
    - [ ] Reviewer 1
    - [ ] Reviewer 2
- Documentation added, function/classes headers clean?
    - [ ] Reviewer 1 check
    - [ ] Reviewer 2 check
- Has the `README.md` been updated, or not required?
    - [ ] Reviewer 1 check
    - [ ] Reviewer 2 check
- Has the `CHANGELOG` been updated?
    - [ ] Reviewer 1 check
    - [ ] Reviewer 2 check
- Have unit tests been added where needed?
    - [ ] Reviewer 1 check
    - [ ] Reviewer 2 check
- Any testing not covered by unit tests was done manually by the reviewers.
    - [ ] Reviewer 1
    - [ ] Reviewer 2
- Version ticks in `package.xml`, `CMakeLists.txt` and/or `setup.py` were made.
    - [ ] Reviewer 1 check
    - [ ] Reviewer 2 check
- Licensing updates were added, or not required.
    - [ ] Reviewer 1 check
    - [ ] Reviewer 2 check
- New dependencies added to `requirements.txt`, `setup.py` and `package.xml` - or: not applicable.
    - [ ] Reviewer 1 check
    - [ ] Reviewer 2 check

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,14 @@
+***
+> **NOTE**: Sunrise robotics has forked the original package and made changes to support their needs.
```
Comment (JenniferBuehler) [inline README.md]: Did you know you can format this nicely :) I only recenlty discovered too. [See here](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts), example [in my PR](https://github.com/Sunrise-Robotics/sunrise_simulation/tree/SUN-865-create-initial-sunrise-simulation).

```
> [!NOTE]
> Useful information that users should know, even when skimming content.
```

**Review by JenniferBuehler — APPROVED** [review]
I built it, it works. Nice!

I have general concerns about how we keep this in sync with upstream (or is our fork so far evolved?) but that's a separate question to this PR. I asked on slack already :)

**Comment by JenniferBuehler** [comment]
No idea why I can't tick the checkmarks on the PR description... maybe it's the recent `apt upgrade` result that I was semi-forced to do and has already given me a bit of headache (with nvidia, whatta surprise)

File: README.md (hunk 1 above)
Comment (fgrcar) [reply README.md]: I also used this feature in sunrise_isaac. Comes out very nice :slightly_smiling_face: 
image

File: abb_hardware_interface/src/abb_hardware_interface.cpp [hunk 2]
```diff
@@ -30,6 +32,13 @@ CallbackReturn ABBSystemHardware::on_init(const hardware_interface::HardwareInfo
     return CallbackReturn::ERROR;
   }
 
+  // Initialize the is_activated_ flag
+  is_activated_ = false;
+
+  // Define the initial joints variable
+  // std::vector<abb::robot::InitialJointValue> initial_joint_values = {};
```
Comment (fgrcar) [inline abb_hardware_interface/src/abb_hardware_interface.cpp:39]: If this is not needed, please remove it.

File: abb_hardware_interface/src/utilities.cpp [hunk 3]
```diff
@@ -42,6 +42,24 @@
 #include <abb_hardware_interface/utilities.hpp>
 #include <stdexcept>
 
+// New includes
```
Comment (fgrcar) [inline abb_hardware_interface/src/utilities.cpp:45]: Gotta love C++, right :exploding_head: :sweat_smile:

File: abb_hardware_interface/src/utilities.cpp [hunk 4]
```diff
@@ -115,6 +133,139 @@ bool verifyStateMachineAddInPresence(const SystemIndicators& system_indicators)
 {
   return system_indicators.addins().state_machine_1_0() || system_indicators.addins().state_machine_1_1();
 }
+
+// This function works as "stand-alone" does not make any use of the RWS managers. That is because
```
Comment (fgrcar) [inline abb_hardware_interface/src/utilities.cpp]: ` works as "stand-alone" AND does not ...`

File: abb_hardware_interface/src/utilities.cpp [hunk 5]
```diff
@@ -115,6 +133,139 @@ bool verifyStateMachineAddInPresence(const SystemIndicators& system_indicators)
...
+// the RWS managers (in their current state) are not compatible with the Omnicore controllers. The
+// primary use of this function is to allow the hardware interface to retrieve the current joint
+// values from the controller before EGM is initialized. This prevents sudden jumps in the joint
+// states which would otherwise occur when starting the hardware interface.
+std::vector<abb::robot::InitialJointValue> getRWSJointsFromController(
```
Comment (fgrcar) [inline abb_hardware_interface/src/utilities.cpp:142]: Damn ... this function makes me sick :joy:

File: abb_hardware_interface/src/utilities.cpp [hunk 6]
```diff
@@ -115,6 +133,139 @@ bool verifyStateMachineAddInPresence(const SystemIndicators& system_indicators)
...
+        // Navigate to the li element with class "ms-jointtarget"
+        Poco::XML::Element* root = doc->documentElement();
+        if (!root) {
+            throw std::runtime_error("Failed to find root element in XML");
+        }
+
+        // Find the li element with class "ms-jointtarget"
+        Poco::XML::Element* body = root->getChildElement("body");
+        if (!body) {
+            throw std::runtime_error("Failed to find body element");
```
Comment (fgrcar) [inline abb_hardware_interface/src/utilities.cpp]: I would add punctuation marks to all those error messages.

**Review by fgrcar — APPROVED** [review]
I can imagine how painful this was :people_hugging:.
Check the comments and you're good to go.

**Comment by fgrcar** [comment]
Btw, I can't tick the checkboxes in the description, so something is wrong with the permissions I guess.

**Pull request merged**